66 proofread documentation: economics and supported proof systems pages (#68)
NilFoundation/proof-market-toolchain: gitbook-docs/market/economics.md

# Economics

> {% hint style="info" %}

Reviewer: I really don't think it should be a callout block. A page shouldn't start with a callout, because a callout is what interrupts the story with some important information. But this is just a normal intro. Hello, this document is about this and that.

This stage may result in a completed deal if proof was provided and verified by the Proof Market
or a restart of the matching if no proof has been provided or proof's verification has failed.
Moreover, some applications require generating proofs in the specified time interval.
In the future, the following parameters will influence the matching algorithm:

* Proof generation time.
  Proof producers provide time bounds _t\_p_ in which they can generate proof for the statement.
  Users can set the desired proof generation time _t\_r_.
  Only proposals with _t\_p_ ≤ _t\_r_ can be matched with the user's request.
* Proof producer's rating.
  Proof producer's rating depends on the rate of completed orders — orders where the proof producer
  has provided correct proof in the requested time interval.

Here are the basic rules of order matching for a request with specified time _t\_r_ and price _c\_r_:

1. Choose proposals with _t\_p_ ≤ _t\_r_ and _c\_p_ ≤ _c\_r_.
2. Sort them by price and proof generation time.
   The user's request defines the priority of these parameters.
3. Apply proof producer's rating to the sorted list.
4. Choose the top proposal from the list.

> ## Requests & proposals resale

Reviewer: I think, it should be "Proof resale" because one cannot sell an order.

  of the original request.
* Unspent tokens are returned to the proof requester.

The total amount of all commissions should be significantly lower than the market price of the request.
This approach allows the proof requester, whose proof will be reused, to recoup proof costs
in the future and maybe even earn something on it.
Similarly, a proof producer can sell their proof more than once and receive a commission
from each such additional sale.

## Registration
### Proof requester

There is no process for proof requesters registration.
Any `=nil;` cluster account can send a request to the marketplace.
The only requirement is that the user's account must have enough funds to pay for the order.

### Proof producer

Account has to register as a Proof Producer to put proposals on the marketplace.
The process includes locking funds (_deposit_).
The number of locked funds reflects the number of orders that the proof producer can process at the moment.

Let _X_ be the number of locked funds by the proof producer.
Then they can process requests on the sum of _X/pf_ at once.
This restriction is related to the penalties policy described in the following sections.

> ## Funds transferring & commissions

Reviewer: Let's not use `&` because it's not a brand or another name.

The only requirement is that the user's account must have enough funds to pay for the order.

### Proof producer

Account has to register as a Proof Producer to put proposals on the marketplace.
The process includes locking funds (_deposit_).
The number of locked funds reflects the number of orders that the proof producer can process at the moment.

Let _X_ be the number of locked funds by the proof producer.
Then they can process requests on the sum of _X/pf_ at once.
This restriction is related to the penalties policy described in the following sections.

## Funds transferring & commissions

When placing a request, the proof requester sets the cost they are willing to pay for
the successfully submitted proof.
If the proof for the request is provided on time, the price indicated in the proposal cost
is deducted from the proof requester's account and credited to the account of the proof producer.

_c_% of the request cost goes as the commission, but not less than _mc_ tokens.
It's split between two parties:

* _app\_dev_ part goes to the Proof Market development team;
* _cir\_dev_ part goes to the circuit developer of the ordered statement.

### Funds withdrawal

> #### From the proof requester's perspective

Reviewer: I like that you converted these wannabe-headings into real ones.

The previous heading is `##`, so this should be `###`, same for the next one.

Author: Funds transferring & commissions was a 2nd level heading, not the previous one
Economics -> Funds transferring & commissions -> Funds withdrawal -> From the proof requester's perspective. Looks like a 4th level heading to me

Reviewer: oh, didn't notice it :)

Each time a new request is added, an amount equal to the request cost is locked on the account.
The locked funds are debited from the account once the request is successfully closed.
If the request is not completed within the specified time interval, the proof requester
can withdraw the request.
If the request is successfully withdrawn, the locked funds will be unlocked.

#### From the proof producer's perspective

At any moment, a proof producer can withdraw funds that are not part of the deposit
for their active proposals.
This means that the proof producer has _pf∗S_ locked funds where _S_ is the sum of all
their proposal orders that are not completed or withdrawn.

## Ratings

Each proof producer has a rating in the system.
Rating impacts many processes, such as matching, proposal placement, and others.
For simplicity, the conscientious fulfillment of obligations increases the rating.
It increases the likelihood of a request matching, while the producer's dishonest work
can lead to decreased ratings and various restrictions.

{% hint style="info" %}
The rating system is in the design/development stage; more details on it will be provided soon**
{% endhint %}

> ## What happens if the proof is not submitted on time

Reviewer: This section has the same topic as "Penalties", so they should be merged. Let's discuss how to do it.

* a limit on the minimum generation time for the producer.

Requests may be unmatched and reassigned to another proof producer if the proof was not submitted on time.

Proof producers don't receive payments for proofs that were not provided in the requested time interval.
However, there's a difference in rating change for when proof was not provided on time
and when it was not provided at all.
Remember that proof producers define their minimum proof generation time, and the matching algorithm
never matches proposals to the requests with _t\_r_ < _t\_p_.

## The delay between the request/proposal submission and the matching

There is no delay between the request/proposal submitting and running the matching algorithm.
As soon as the request/proposal is submitted, the Proof Market tries to match it with
other existing orders.

## Penalties

There are two types of penalties for proof producers:

* Decreasing the rating.
  It applies when proof was not provided on time or was not provided at all.
* A fine on the locked funds.
  When verification fails for the provided proof, the proof producer pays _pf \* c_ tokens as a fine,
  where _c_ is the order's cost.

>   This is considered malicious behavior from the proof producer side.

Reviewer: Unclear what's *this* in the last sentence. The fine?

Overall, this whole block is reversed. We group it by penalties and then explain what can cause such penalties.

What if we talk about misbehavior first?

* If a proof producer provides the proof out of time or doesn't provide it at all, their rating will decrease.
* If a proof producer provides proof that doesn't verify, a fine of _pf \* c_ tokens will be subtracted from their deposit, where _c_ is the order's cost.

# Economics

> {% hint style="info" %}

Reviewer: yep, definitely not danger

> ### Disclaimer

Reviewer: Headings go from level 1 straight to level 3. That's not right. Please move the headings up a level.

| _app\_dev_ | 0.7   | application developers part of the fee |
| _cir\_dev_ | 0.3   | circuit developers part of the fee     |

### Basic workflow

1. A proof buyer sends a request with a desired price _c\_b_ to the market.
2. The Proof Market locks _c\_b_ tokens from the buyer's account.

> 3. A proof producer sends a proposal to the market with a price _c\_a <= c\_b_.

Reviewer: as far as I know, multiple producers can send a proposal, and the market will select one

4. The Proof Market matches the request with the proposal.

> 5. The proof producers generate a proof and send it to the market.

Reviewer: won't there be a single particular producer that generates the proof?

6. The Proof Market verifies proof and pays _c\_b - commission_ tokens to the producer.

> 7. The proof buyer takes their proof and uses it.

Reviewer: Now that we've renamed bids and asks, the variables should be renamed as well. cost_ask → cost_request (c_r).

Reviewer: same with time_ask and time_bid

Author: @NickVolynkin 
> cost_ask → cost_request (c_r)

You mean cost_ask → cost_proposal? And request instead of bid

Reviewer: Yes!

### Orders status

Requests (orders for buying proofs) and proposals (orders for selling proofs) have the following statuses:

> * _Created_: order was sent to the marketplace and was not yet matched,

Reviewer: which of the orders: order for buying or order for selling?

I think, it's an ambiguous term, and we should avoid using it.

Author: All orders, I suppose? According to the next line

Reviewer: How can there be no buyer for a proposal if the proposal is made for a particular request?

Or can proof producers post long-term proposals for a circuit? Like, ready to prove any mina-state-proof requests, 5 bucks each, shop open 24/7?

Reviewer: Okay, we need to put state diagrams for both requests and proposals. It's definitely a separate task.

  meaning there is no buyer for a proposal or no proof producer for a request.
* _Processing_: order was matched, and the Proof Market is waiting for the proof generator
  to provide the proof.
* _Completed_: the proof was generated and verified by Proof Market.
* _Withdrawn_: the order was canceled before matching, a wrong proof was submitted, or

>   no proofs were submitted in the requested time interval.

Reviewer: Say, the proof producer submitted bad proof. The PM tried to verify it and saw that it doesn't verify. Shouldn't it find a different proof?

### Request/proposal matching process

The Proof Market provides a more complex behavior than one-hop trading.
Each deal includes the proof generation stage.
This stage may result in a completed deal if proof was provided and verified by the Proof Market
or a restart of the matching if no proof has been provided or proof's verification has failed.
Moreover, some applications require generating proofs in the specified time interval.
In the future, the following parameters will influence the matching algorithm:

* Proof generation time.
Proof producers provide time bounds _t\_a_ in which they can generate proof for the statement.
Users can set the desired proof generation time _t\_b_.

> Only proposals with the _t\_a_ ≤ _t\_b_ can be matched with the user’s request.

Reviewer: remember about the indentation in lists

>   Only proposals with the _t\_p_ ≤ _t\_r_ can be matched with the user’s request.

Reviewer: let the documentation engine make curly braces for you.
```suggestion
  Only proposals with the _t\_p_ ≤ _t\_r_ can be matched with the user's request.
```

> * Proof producers rating.

Reviewer: ```suggestion
* Proof producer's rating.
```

or 

```suggestion
* Proof producer ratings.
```

  Proof producer's rating depends on the rate of completed orders — orders where the proof producer
  has provided correct proof in the requested time interval.

Here are the basic rules of order matching for a request with specified time _t\_r_ and price _c\_r_:

1. Choose proposals with _t\_p_ ≤ _t\_r_ and _c\_p_ ≤ _c\_r_.
2. Sort them by price and proof generation time.

>    The user’s request defines the priority of these parameters.

Reviewer: same about curly braces

3. Apply proof producer's rating to the sorted list.
4. Choose the top proposal from the list.

### Requests & proposals resale

There may be situations when an equivalent request is already in the database.
Two requests are considered equivalent if they match in the **statement** and **public\_input** fields.
Upon receiving a new request, the application tries to find an equivalent request in the system.
This is what happens if such a request is found and there's already proof for it:

* The new request immediately goes into the **completed** state.

> * The proof requester pays commissions to the Proof Market, the proof producer who generated the proof

Reviewer: And how about the commission to the circuit developer (remember the 0.7/0.3 proportion).

> for the existing request, and the owner of that request.

Reviewer: who is the owner of the request? Some new role?

Author: How about “the proof buyer/requester of the original request”?
Btw, in this article we have terms “proof buyer” and “proof requester” for the same entity, should I fix that?

* Unspent tokens are returned to the proof requester.

The total amount of all commissions should be significantly lower than the market price of the request.

> This approach allows the request owner, whose result will be reused, to recoup request costs

Reviewer: We've got some ambiguity in the terms here.

There are two types of requests, so a request owner can be both a proof requester and a proof producer. Let's avoid using the term "request owner". Instead, use "proof requester" and "proof producer".

```suggestion
This approach allows the proof requester, whose proof will be reused, to recoup proof costs
```

in the future and maybe even earn something on it.
Similarly, a proof producer can sell their proof more than once and receive a commission
from each such additional sale.

### Registration

#### Proof requester

There is no process for proof requesters registration.
Any `=nil;` cluster account can send a request to the marketplace.
The only requirement is that the user's account must have enough funds to pay for the order.

#### Proof producer

Account has to register as a Proof Producer to put proposals on the marketplace.
The process includes locking funds (_deposit_).
The number of locked funds reflects the number of orders that the proof producer can process at the moment.

Let _X_ be the number of locked funds by the proof producer.
Then they can process requests on the sum of _X/pf_ at once.
This restriction is related to the penalties policy described in the following sections.

### Funds transferring & commissions

When placing a request, the customer sets the cost they are willing to pay for
the successfully submitted proof.
If the proof for the request is provided on time, the price indicated in the proposal cost

> is deducted from the customer's account and credited to the account of the proof producer.

Reviewer: It's a customer, but the role is named a "proof requester".
```suggestion
is deducted from the proof requesters's account and credited to the account of the proof producer.
```

_c_% of the request cost goes as the commission, but not less than _mc_ tokens.
It's split between two parties:

* _app\_dev_ part goes to the Proof Market development team;
* _cir\_dev_ part goes to the circuit developer of the ordered statement.

> #### Funds withdraw

Reviewer: ```suggestion
#### Funds withdrawal
```

**From the proof requester's perspective:**

Each time a new request is added, an amount equal to the request cost is locked on the account.
The locked funds are debited from the account once the request is successfully closed.
If the request is not completed within the specified time interval, the customer can withdraw the request.
If the request is successfully withdrawn, the locked funds will be unlocked.

**From the proof producer's perspective:**

At any moment, a proof producer can withdraw funds that are not part of the deposit
for their active proposals.
This means that the proof producer has _pf∗S_ locked funds where _S_ is the sum of all
their proposal orders that are not completed or withdrawn.

### Ratings

Each proof producer has a rating in the system.

> Rating impacts many processes, such as matching, proposal placement, etc.

Reviewer: etc. -> and others.
